### PR TITLE
Disk hotplug for VMs

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [inline lxd/instance/drivers/qmp/commands.go:266]: You can just call `m.run("blockev-del", blockDevDel, nil)` here, no need to test for the result.

Comment (monstermunchkin) [reply lxd/instance/drivers/qmp/commands.go:266]: Right. Also, there's a typo which needs to be fixed :)

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [inline lxd/instance/drivers/qmp/commands.go:305]: I'm not a fan of this sort of arbitrary re-try logic in lower level libraries (such as this one the `qmp` package), as we can't know the situation that the caller is calling us from, and whether it is appropriate to retry or not, and for how long/how many times/how often.

Have you identified when this situation occurs? Does it only happen during hot-unplug of a running device, but not during device removal of a VM that is shutting down?

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1786]: A comment here explaining what you're splitting up would be useful to the reader.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1788]: I'm a bit concerned by this, as according to the comment on the `DevPaths()` function, d.devPaths is only fully populated with the device paths for all devices during instance startup. However presumably the hot plug feature will be called both at start up and after start up - at which point the d.devPaths will be empty, and you'll potentially end up replacing formerly allowed device paths added to the apparmor profile during start with just the hot plugged disk device path. Have I read this right?

Also a comment here explaining what is going on would be useful as it required me to read around the instance code base to figure out what was going on here.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1854]: Same concern about d.devPaths here as above, if this is run when the VM is already running (i.e hot-unplug) won't we end up removing all previously allowed devPaths?

File: lxd/instance/drivers/qmp/commands.go
Comment (monstermunchkin) [reply lxd/instance/drivers/qmp/commands.go:305]: > Have you identified when this situation occurs? Does it only happen during hot-unplug of a running device, but not during device removal of a VM that is shutting down?

I believe it's a timing issue. It happens when a disk is hot-unplugged from a running VM. This code isn't called when a VM is shutting down.

Comment (tomponline) [reply lxd/instance/drivers/qmp/commands.go:305]: In that case the retry logic should be pushed to where its situationally relevant (i.e `deviceDetachBlockDevice`) and out of the `qmp` package. You may want to achieve this by returning a specific error var or error type that is recognisable as a retriable error.

Comment (tomponline) [reply lxd/instance/drivers/qmp/commands.go:305]: Just like the wait logic in `deviceDetachNIC`.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:1788]: > However presumably the hot plug feature will be called both at start up and after start up - at which point the d.devPaths will be empty, and you'll potentially end up replacing formerly allowed device paths added to the apparmor profile during start with just the hot plugged disk device path. Have I read this right?

The hot plug feature is only called when a device is being hot-plugged, i.e. not at VM startup.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:1854]: We iterate through devPaths and only skip the entry of the disk which is to be removed. We don't remove all previously allowed devPaths.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1788]: Ideally we would have this feature behave like the NIC hot-plug feature, which looking at `addNetDevConfig()` (which is the function that then calls `m.AddNIC()` on the `qmp` package, its called twice, at start up and at hot-plug time.
This reason I went this route rather than adding device config to the qemu.conf file, is to minimise the variations in device management. Effectively making the QMP add/remove approach the *only* way we add/remove NICs.

This is also preferred by the upstream QEMU team who are looking at phasing out the qemu.conf altogether, and we've seen a few breakages in parsing that file. Whereas using QMP is the recommended way to manage devices.

So for this PR I would expect to be seeing the disk devices (or at least the ones that can be unplugged) being removed from static config file and switching to use QMP at start up to add them.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1854]: Yes i see that, but my understanding is that devPaths itself will be empty any other time than start up. So effectively this would remove start-up time devPaths from the apparmor profile.

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1854]: https://github.com/lxc/lxd/blob/57e0dbc26c2bde3ce2fc1ce33e5528fd16ef489b/lxd/instance/drivers/driver_common.go#L94-L97

Which makes sense, because we don't keep the `Instance` var instance around during the VM lifetime, a  new `Instance` var instance is created when updating a VM.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1788]: > OK, I'll rework the code to use QMP for all disks except maybe the root disk.

Thanks, if you can use it for the root disk even better, as it avoids the variety of code paths we have to support.

**Review by tomponline — CHANGES_REQUESTED** [review]
See comments above.

**Comment by tomponline** [comment]
@monstermunchkin I noticed you've pushed some changes, did you want another review yet. If not no worries :)

**Comment by monstermunchkin** [comment]
> @monstermunchkin I noticed you've pushed some changes, did you want another review yet? If not no worries :)

Sure, why not :) The changes should cover all your comments.

**Comment by tomponline** [comment]
> > @monstermunchkin I noticed you've pushed some changes, did you want another review yet? If not no worries :)
> 
> Sure, why not :) The changes should cover all your comments.

OK cool will do, let me know if in the future if you're after another review in case I forget to check. Cheers

File: lxd/device/device_common.go
Comment (tomponline) [inline lxd/device/device_common.go:97]: As this is very QEMU specific and potentially error prone because the host path returned from Start() may be dynamic based on run-time configuration, how about storing the device paths returned from Start() in a json file in the instance's volume generated by the qemu driver, which can then be later be read back & edited, and used to regenerate the AppArmor profile?

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:3237]: Is there a reason why the call to `AddBlockDevice` in `deviceAttachBlockDevice` is missing a lot of the caching and other logic from this function? Would it make sense to centralize this logic so its used for both start time and hot-plug time?

**Review by tomponline — CHANGES_REQUESTED** [review]

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go]: I'll keep in this PR.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:1820]: This should actually be
```go
monHook, err := m.addDriveConfig(...)
```
which makes a lot more sense. That's how it's done for NICs as well.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:3237]: As stated above, this should be `monHook, err := d.addDriveConfig(...)`.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:3209]: Our current limit for device names is 63 characters.

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:3209]: OK, I think we should just detect the specific scenario here and error out. We'll also need to take into account the `qemuBlockDevIDPrefix`

Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:3209]: This PR prefixes block devices with `lxd_` which means our limit would be 27 characters.

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:3209]: Yep

File: lxd/device/device_common.go
Comment (monstermunchkin) [reply lxd/device/device_common.go:97]: We can do that.

Comment (monstermunchkin) [reply lxd/device/device_common.go:97]: There's one small issue if we removed `DevPath`. When a VM is running and LXD is restarted (VM is still running), there is no way for the newly started LXD server to know `devPaths` if the JSON file doesn't exist yet. If we then hotplug a disk, `devPaths` (and the JSON file) will only contain the disk path which is incorrect. A restart of the VM obviously fixes that.

Is this a problem we should consider?

Comment (tomponline) [reply lxd/device/device_common.go:97]: If its been deleted then we can expect problems but that seems unlikely. You can detect that and error before reloading aparmor though.

Comment (monstermunchkin) [reply lxd/device/device_common.go]: This corner case would also happen after an update, so we shouldn't return an error in this case.

Comment (tomponline) [reply lxd/device/device_common.go]: What do you mean? You mean the first time lxd is updated to a version that supports this feature? If so then we have to error as the alternative is wiping out the apparmor Dev paths which isn't an option as it would break a running vm.

Comment (monstermunchkin) [reply lxd/device/device_common.go]: Yes, that's what I meant. I would've liked a solution which works in the update case as well. Anyway, it now uses a `devpaths.json` which is required when updating devices.

**Comment by monstermunchkin** [comment]
@tomponline feel free to take another look.

File: lxd/device/device_common.go
Comment (tomponline) [reply lxd/device/device_common.go]: I see what you would prefer such an approach. And I'm not dead set against it at all. However my concern is that in its original form as a sort of "bolt-on" concept to the `device` type that it would easily get forgotten or broken when adding new devices or changing the behavior of existing ones, leading to strange bugs where editing one disk would then break another.

I think if we were going to use `DevPath()` as a concept then we would want to formalise the contract it gives to the instance drivers. By that I mean that we should then stop passing the device path via the RunConf returned from `Start()` and `Stop()` and instead update the instance drivers to call the device's `DevPath()` function after calling `Start()` or `Stop()`. This way should we ever find ourselves breaking that contract (i.e not being able to provide the `DevPath()` for a disk device outside of the `Start()` function) then we would immediately know about it as the instance wouldn't be able to start.

Also as `DevPath()` is rather disk specific, I suggest that rather than adding it to the main `Device` interface, we use interface-smuggling to create a new interface called `DeviceDisk` that provides `DevPath()` and that way one would be forced to implement it before the instance could use the device.

What do you think?

Comment (monstermunchkin) [reply lxd/device/device_common.go]: I don't see why we would need to remove the device path from RunConf. That would just cause a lot of unnecessary code changes which are unrelated to this PR. The only reason I want a `DevPath` function is for when `devpaths.json` doesn't exist.

How about we just create a new interface `Disk` (as opposed to `DeviceDisk` as golint doesn't like that) for now which implements `DevPath`, and is implemented by disk and usb.

Comment (tomponline) [reply lxd/device/device_common.go]: > The only reason I want a `DevPath` function is for when `devpaths.json` doesn't exist.

We should not need the devpaths.json feature at all if we are implementing `DevPath` in the device.

However my concern is that `DevPath()` is only used populating the apparmor profile and isn't used for actually setting up the VM config. And because we are decoupling the returned `RunConf` (which is used in the QEMU config) from the `DevPath()` value (which is used for AppArmor profile) this leads to multiple potential problems down the line:

1. Firstly, there is the possibility that we may introduce a bug where the value of `DevPath()` and the dev path in `RunConf` differ for a device. We shouldn't have 2 sources of "truth" for a disk device's source device path.
2. By introducing `DevPath()` for AppArmor profile usage we are subtly changing the contract between instance and disk device drivers. Previously the contract was effectively that by calling device.Start() the device could return a start-time disk source device path in `RunConf` and the instance driver would use it. But now we are changing the contract to require the disk device to be able to provide the disk source device path at a non-start time (i.e `DevPath()` can be called at any time - and infact we rely on that because we are calling it for all disk devices at update time when another disk is hot-plugged).

Now this change of contract is fine (I'm assuming you've checked all scenarios of disk devices and all of them can provide a non-start time disk source device via `DevPath()`), but we should formalise it so that we don't accidentally introduce a change in the future that breaks that contract (perhaps a disk device mode that only knows the source of the disk device at start time due to some dynamic allocation of block devices) such that a non-start time call of `DevPath()` cannot be satified. In that case if we were not using `DevPath()` as the single source of truth for disk device setup then we likely wouldn't realise its a problem until we start getting strange issues due to apparmor profiles being changed for one disk device and removing the start-time added dev path for the dynamic disk device.

Also the way its was implemented by having all device types implement an empty `DevPath()` if not explicitly implemented meant that this too could be forgotten, whereas not have it implemented in the common driver and instead using interface smuggling will force the developer to remeber that `DevPath()` needs implementing.

Comment (tomponline) [reply lxd/device/device_common.go]: To help break the changes down, it would be good to have this `DevPaths` change as a separate PR, and this would also help to catch regressions.

Comment (monstermunchkin) [reply lxd/device/device_common.go]: > Firstly, there is the possibility that we may introduce a bug where the value of DevPath() and the dev path in RunConf differ for a device. We shouldn't have 2 sources of "truth" for a disk device's source device path.

They will most likely differ, as in `DevPath` we only care about block devices. Everything else is ignored.

Comment (tomponline) [reply lxd/device/device_common.go]: I'm not sure I follow you I'm afraid. Are they not representing the same thing? (If not then `DevPath(`  should be named something else to avoid confusion).

**Comment by monstermunchkin** [comment]
Do not merge yet. There's an issue with the qmp fds which needs fixing. I discovered it while working on the USB hotplug support.

**Comment by tomponline** [comment]
> Do not merge yet. There's an issue with the qmp fds which needs fixing. I discovered it while working on the USB hotplug support.

OK.

**Comment by monstermunchkin** [comment]
The issue is fixed. The problem was that `SendFileWithFDSet` didn't store the JSON data correctly as the struct was incorrect. This lead to the returned FD set ID always being `0`. I fixed the struct, and now the correct FD set ID is returned.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1783]: > This function only operates on a single "mount", so accepting a slice of `mounts` and then ignore the other potential entries is problematic. Please can you update it to just accept a single `deviceConfig.MountEntryItem` item.

@monstermunchkin this point is not addresses yet. ^

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go]: This logic has the same problem as the original proposal of the `DevPath()` function, i.e it is assuming that for all disk types the `source` property available at non-start time is the same source path that was returned by the device's start-time RunConfig.

Now that we are recording into the json file the actual source returned from the device's start-time RunConfig we need to remove that value from the AppArmor profile, which is not necessarily the one in `source`.

Instead I suggest expanding the json file struct to look a bit like `NICConfig` struct, i.e containing `DeviceName` and then `SourcePath`. This way you can match on `DeviceName` and not include the associated `SourcePath` in the apparmor profile.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:2726]: > No need for this line here.

Still needs addressing.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:3166]: > Please could you expand on this? What permission issue did you encounter?

Can you answer this one? thanks

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go]: Instead of introducing a new struct, I'd go with a `map[string]string`, the key being the device name and the value the source path.

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go]: I would suggest using a new struct type as it allows for easier expandibility in the future should we need to add more fields. Also it has the added benefit that we can then easily see where that type is used, as compared to map[string]string which is used all over the place.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:1816]: You mean `lxd/instance/drivers/qemu: Use fdset for block device` and `lxd/instance/drivers/qmp: Add FD set commands` should be one commit? I'd prefer to keep them separate as they both touch different files/modules.

Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:1824]: No sorry I explained myself badly. I meant I was expecting to see just one commit for adding `deviceDetachBlockDevice` function.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:3166]: Actually, this works now. It didn't in a previous version of the PR, but now it does. Yay :tada:

Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go]: This is fixed now.

**Comment by tomponline** [comment]
@monstermunchkin @stgraber just suggested that perhaps the whole AppArmor dev paths issues is now moot given that you're now passing everything using FDs. Can you check if the AppArmor dev paths concept is even needed anymore? Thanks

**Comment by monstermunchkin** [comment]
> @monstermunchkin @stgraber just suggested that perhaps the whole AppArmor dev paths issues is now moot given that you're now passing everything using FDs. Can you check if the AppArmor dev paths concept is even needed anymore? Thanks

I removed that code as it works without it.

**Comment by tomponline** [comment]
> > @monstermunchkin @stgraber just suggested that perhaps the whole AppArmor dev paths issues is now moot given that you're now passing everything using FDs. Can you check if the AppArmor dev paths concept is even needed anymore? Thanks
> 
> I removed that code as it works without it.

Excellent, thanks for the tip @stgraber, I missed that when we added FD support the previously required dev paths for apparmor went away.

**Comment by monstermunchkin** [comment]
I haven't removed the entire `devPaths` logic as USB devices still rely on it. Once USB hot-plugging is implemented, we might be able to remove it entirely.

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go:2777]: That's what I thought, too. However, the root device is not first and therefore this is needed.

File: lxd/instance/drivers/qmp/commands.go
Comment (monstermunchkin) [reply lxd/instance/drivers/qmp/commands.go:18]: Taken from the QMP docs:

    opaque: string (optional)
        A free-form string that can be used to describe the fd.

File: lxd/instance/drivers/qmp/commands.go
Comment (monstermunchkin) [reply lxd/instance/drivers/qmp/commands.go:423]: I can make it consistently return `ErrNotFound` (or `http.StatusNotFound`) in all cases.

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [reply lxd/instance/drivers/qmp/commands.go:18]: Cool thanks, its a terrible name, "description" would have been better, but oh well :)

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [reply lxd/instance/drivers/qmp/commands.go:423]: Unless there's a specific reason we need to know it wasn't there when it was removed I think we can just have to return nil if it was not found (i.e the end result is the same, it was removed).

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:2777]: OK I think we need to look into why that is and fix it rather than introduce a specific workaround here.

File: lxd/instance/drivers/qmp/commands.go
Comment (monstermunchkin) [reply lxd/instance/drivers/qmp/commands.go:19]: I totally agree :)

File: lxd/instance/drivers/driver_qemu.go
Comment (monstermunchkin) [reply lxd/instance/drivers/driver_qemu.go]: Done

File: lxd/device/config/devices_sort.go
Comment (tomponline) [inline lxd/device/config/devices_sort.go:61]: @monstermunchkin did the previous one not work because of disk devices with no `path`?

Comment (monstermunchkin) [reply lxd/device/config/devices_sort.go:61]: Yes

Comment (tomponline) [reply lxd/device/config/devices_sort.go:61]: Cool, nice we fixed that then.

**Comment by monstermunchkin** [comment]
jenkins: test this please

**Comment by monstermunchkin** [comment]
@tomponline @stgraber All comments were addressed. Should be good to go.

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [inline lxd/instance/drivers/qmp/commands.go:469]: deviceID arg not used, can it be removed?

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [inline lxd/instance/drivers/qmp/commands.go:399]: We can just return nil here right? (the device isn't present, so its removed already, nothing to do).

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [inline lxd/instance/drivers/qmp/commands.go:424]: We can just return nil here right? (the device isn't present, so its removed already, nothing to do).

File: lxd/instance/drivers/qmp/commands.go
Comment (tomponline) [inline lxd/instance/drivers/qmp/commands.go:395]: Nice!

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:2002]: If `RemoveDevice` returned nil if not found this would be simpler.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1842]: No need for this if `RemoveBlockDevice` returns nil if not found.

File: lxd/instance/drivers/driver_qemu.go
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:1830]: This could be simpler if `RemoveDevice` returned nil if not found.

**Review by tomponline — CHANGES_REQUESTED** [review]
LGTM thanks, just a couple of minor things around the device removal errors and one redundant arg and this is good to go.

Do you have a PR I can look at for the lxc-ci tests yet? Thanks

File: lxd/instance/drivers/qmp/commands.go
Comment (monstermunchkin) [reply lxd/instance/drivers/qmp/commands.go:399]: I believe that's what I had and you suggested to do this.

Comment (tomponline) [reply lxd/instance/drivers/qmp/commands.go:399]: I did suggest replacing the ErrNotFound with this that you originally had, but after looking at it more I suggested just returning nil if not found.

**Comment by monstermunchkin** [comment]
@tomponline good to go

**Review by tomponline — APPROVED** [review]
LGTM thanks!

**Pull request merged**
